crypto/params.c

```diff
@@ -0,0 +1,254 @@
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_SET_DOUBLE, CRYPTO_R_PARAM_NOT_FOUND);
+        ERR_add_error_data(2, "param name=", key);
+        return 0;
+    }
+
+    /* Type safety. */
+    width = OSSL_PARAM_get_width(OSSL_PARAM_DOUBLE);
+    if (p->return_size != NULL)
+        *p->return_size = width;
+    if (p->data_type != OSSL_PARAM_DOUBLE || p->buffer_size != width) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_SET_DOUBLE, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+    dp = (double *)p->buffer;
+    *dp = val;
+    return 1;
+}
+
+/*
+ * Same comment, but the width is always sizeof(void *)
+ */
+int OSSL_PARAM_get_pointer(const OSSL_PARAM *p, const char *key, void **val)
+{
+    size_t width;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_POINTER, CRYPTO_R_PARAM_NOT_FOUND);
+        ERR_add_error_data(2, "param name=", key);
+        return 0;
+    }
+
+    /* Type safety. */
+    width = OSSL_PARAM_get_width(OSSL_PARAM_POINTER);
+    if (p->data_type != OSSL_PARAM_POINTER || p->buffer_size != width) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_POINTER, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+
+    memcpy(val, p->buffer, width);
+    return 1;
+}
+
+int OSSL_PARAM_set_pointer(const OSSL_PARAM *p, const char *key, void *val)
+{
+    size_t width;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_SET_POINTER, CRYPTO_R_PARAM_NOT_FOUND);
+        ERR_add_error_data(2, "param name=", key);
+        return 0;
+    }
+
+    /* Type safety. */
+    width = OSSL_PARAM_get_width(OSSL_PARAM_POINTER);
+    if (p->return_size != NULL)
+        *p->return_size = width;
+    if (p->data_type != OSSL_PARAM_POINTER || p->buffer_size != width) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_SET_POINTER, CRYPTO_R_TYPE_MISMATCH);
+        return 0;
+    }
+    memcpy(p->buffer, val, width);
+    return 1;
+}
+
+
+BIGNUM *OSSL_PARAM_get_bignum(const OSSL_PARAM *p, const char *key)
+{
+    BIGNUM *b = NULL;
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_BIGNUM, CRYPTO_R_PARAM_NOT_FOUND);
+        ERR_add_error_data(2, "param name=", key);
+        goto end;
+    }
+
+    /* Type safety. */
+    if (p->data_type != OSSL_PARAM_BIGNUM) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_GET_BIGNUM, CRYPTO_R_TYPE_MISMATCH);
+        goto end;
+    }
+
+    b = BN_native2bn(p->buffer, (int)p->bn_size, NULL);
+end:
+    return b;
+}
+
+int OSSL_PARAM_set_bignum(OSSL_PARAM *p, const char *key, const BIGNUM *val)
+{
+    const size_t bytes = (size_t)BN_num_bytes(val);
+    int r;
+
+
+    if ((p = OSSL_PARAM_locate(p, key)) == NULL) {
+        CRYPTOerr(CRYPTO_F_OSSL_PARAM_SET_BIGNUM, CRYPTO_R_PARAM_NOT_FOUND);
+        ERR_add_error_data(2, "param name=", key);
+        return 0;
+    }
+
+    /* Type safety. */
+    if (p->data_type != OSSL_PARAM_BIGNUM
+            || BN_is_negative(val)
```

Comment:
So we can only handle positive bignums? Not sure that's a good idea.

Reply:
You need to look at `BN_bn2bin` and friends, and especially this from the manual:

> **BN_bn2bin()** converts the absolute value of **a** into big-endian form and stores it at **to**. **to** must point to BN_num_bytes(**a**) bytes of memory.

I think that #8400 handles it by only supporting bignums as `OSSL_PARAM_UNSIGNED_INTEGER`

So really, if we want to be able to represent negative numbers too, we need to add supporting functions.